src/server/tiering/disk_storage.cc

```diff
@@ -8,12 +8,55 @@
 #include "base/logging.h"
 #include "server/error.h"
 #include "server/tiering/common.h"
+#include "util/fibers/uring_proactor.h"
 
 namespace dfly::tiering {
 
+using namespace ::util::fb2;
+
+namespace {
+
+UringBuf AllocateTmpBuf(size_t size) {
+  size = (size + kPageSize - 1) / kPageSize * kPageSize;
+  uint8_t* buf = new (std::align_val_t(kPageSize)) uint8_t[size];
+  return UringBuf{{buf, size}, std::nullopt};
+}
+
+void DestroyTmpBuf(UringBuf buf) {
+  DCHECK(!buf.buf_idx);
+  ::operator delete[](buf.bytes.data(), std::align_val_t(kPageSize));
+}
+
+UringBuf PrepareBuf(size_t size) {
+  DCHECK_EQ(ProactorBase::me()->GetKind(), ProactorBase::IOURING);
+  auto* up = static_cast<UringProactor*>(ProactorBase::me());
+
+  UringBuf buf;
+  if (auto borrowed = up->RequestBuffer(size); borrowed)
+    return *borrowed;
+  else
+    return AllocateTmpBuf(size);
+}
+
+void ReturnBuf(UringBuf buf) {
+  DCHECK_EQ(ProactorBase::me()->GetKind(), ProactorBase::IOURING);
+  auto* up = static_cast<UringProactor*>(ProactorBase::me());
+
+  if (buf.buf_idx)
+    up->ReturnBuffer(buf);
+  else
+    DestroyTmpBuf(buf);
+}
+
+}  // anonymous namespace
+
 std::error_code DiskStorage::Open(std::string_view path) {
   RETURN_ON_ERR(io_mgr_.Open(path));
   alloc_.AddStorage(0, io_mgr_.Span());
+
+  auto* up = static_cast<UringProactor*>(ProactorBase::me());
+  up->RegisterBuffers(512_KB);
+
```

Comment:
Who's in charge of doing this / deciding on the size?

Reply:
either constexpr or a flag with a sane default